bayes_kit/drghmc.py

```diff
@@ -0,0 +1,437 @@
+                    )
+            return steps
+
+        else:
+            raise TypeError("steps must be None, int, list, or callable")
+
+    def __iter__(self) -> Iterator[DrawAndLogP]:
+        """Use the DrGhmcDiag sampler as an iterator.
+
+        Returns:
+            Iterator[DrawAndLogP]: Iterator of draws from DrGhmcDiag sampler
+        """
+        return self
+
+    def __next__(self) -> DrawAndLogP:
+        """Yields next draw from DrGhmcDiag iterator.
+
+        Yields:
+            DrawAndLogP: Tuple of (sample, log density)
+        """
+        return self.sample()
+
+    def joint_logp(self, theta: VectorType, rho: VectorType) -> float:
+        """Joint log density of sample (theta, rho) under the cannonical distribution.
+
+        Args:
+            theta: position
+            rho: momentum
+
+        Returns:
+            float: joint log density
+        """
+        adj: float = 0.5 * np.dot(rho, self._metric * rho)
+        return self._model.log_density(theta) - adj
+
+    def leapfrog(
+        self,
+        theta: VectorType,
+        rho: VectorType,
+        stepsize: float,
+        steps: int,
+    ) -> tuple[VectorType, VectorType]:
+        """Propose new sample (theta_prop, rho_prop) with leapfrog integrator.
+
+        Discretize and solve Hamilton's equations using the leapfrog integrator to
+        deterministically propose a new sample (theta_prop, rho_prop) from the current
+        sample (theta, rho).
+
+        Copy theta because numpy's += operator mutates the original array instead of
+        creating a new one.
+
+        Initialize rho_mid by going backwards half a step so that the first full-step
+        inside the loop brings rho_mid up to +1/2 steps.
+
+        Args:
+            theta: position
+            rho: momentum
+            stepsize: stepsize in each leapfrog step
+            steps: number of leapfrog steps
+
+        Returns:
+            tuple[VectorType, VectorType]: proposed sample
+        """
+        theta = np.array(theta, copy=True)
+        _, grad = self._model.log_density_gradient(theta)
+        rho_mid = rho - 0.5 * stepsize * np.multiply(self._metric, grad).squeeze()
+        for _ in range(steps):
+            rho_mid += stepsize * np.multiply(self._metric, grad).squeeze()
+            theta += stepsize * rho_mid
+            _, grad = self._model.log_density_gradient(theta)
+        rho = rho_mid + 0.5 * stepsize * np.multiply(self._metric, grad).squeeze()
+        return (theta, rho)
+
+    def _prob_of_delayed_rejection(self, hastings: float) -> float:
+        """Log probability of making another proposal upon rejection.
+
+        To reduce average cost per iteration for DRHMC, make the delayed rejections
+        probabilistic, such that a subsequent proposal is not mandatory upon rejection.
+
+        Probabilistic delayed rejection proposes a new sample with some probability
```

Comment:
At this point either commit fully to waving your hands and make this simpler or instead just write out the rule, which is quite simple.